Sources/_StringProcessing/Engine/InstPayload.swift

```diff
@@ -227,11 +226,64 @@ extension Instruction.Payload {
     interpret()
   }
 
-  init(assertion: AssertionFunctionRegister) {
-    self.init(assertion)
-  }
-  var assertion: AssertionFunctionRegister {
-    interpret()
+  var _assertionKindMask: UInt64 { ~0xFFF0_0000_0000_0000 }
+  init(assertion: AST.Atom.AssertionKind,
+       _ anchorsMatchNewlines: Bool,
+       _ usesSimpleUnicodeBoundaries: Bool,
+       _ usesASCIIWord: Bool,
+       _ semanticLevel: MatchingOptions.SemanticLevel
+  ) {
+    // 4 bits of options
+    let anchorBit: UInt64 = anchorsMatchNewlines ? (1 << 55) : 0
+    let boundaryBit: UInt64 = usesSimpleUnicodeBoundaries ? (1 << 54) : 0
+    let strictBit: UInt64 = usesASCIIWord ? (1 << 53) : 0
+    let semanticLevelBit: UInt64 = semanticLevel == .unicodeScalar ? (1 << 52) : 0
+    let optionsBits: UInt64 = anchorBit + boundaryBit + strictBit + semanticLevelBit
+
+    // 4 bits for the assertion kind
+    // Future work: Optimize this layout
+    let kind: UInt64
+    switch assertion {
+    case .endOfLine: kind = 0
+    case .endOfSubject: kind = 1
+    case .endOfSubjectBeforeNewline: kind = 2
+    case .firstMatchingPositionInSubject: kind = 3
+    case .notTextSegment: kind = 4
+    case .notWordBoundary: kind = 5
+    case .resetStartOfMatch: kind = 6
+    case .startOfLine: kind = 7
+    case .startOfSubject: kind = 8
+    case .textSegment: kind = 9
+    case .wordBoundary: kind = 10
+    }
+    self.init(rawValue: kind + optionsBits)
```

Comment:
I think it would clarify things to encapsulate this logic into a struct. That is the struct would store the `UInt64` and its methods would get/set the bits and kinds. Then the instruction's payload is just the struct's raw value (which we assert doesn't have the opcode bits set at all)